### ISSUE TITLE
Build instructions: error: ‘getGlobalContext’ was not declared in this scope

### ISSUE THREAD

**Issue by mewmew** [revng/revamb#1]
Hej!

What a most wonderful project you have been working on! Thank you Alessandro for sharing it with the public, and thanks for the talk you gave at the last LLVM Developers' Meeting.

I wanted to take rev.ng out for a spin and tried to compile it on my laptop, but ran in to a couple of build errors.

```
u@x220 ~> pacman -Q llvm clang qemu
llvm 3.9.1-2
clang 3.9.1-2
qemu 2.8.0-2
```

```
[ 25%] Building CXX object CMakeFiles/revamb-dump.dir/dump.cpp.o
/home/u/Desktop/revng_play/revamb/dump.cpp: In function ‘int main(int, const char**)’:
/home/u/Desktop/revng_play/revamb/dump.cpp:143:43: error: ‘getGlobalContext’ was not declared in this scope
   LLVMContext &Context = getGlobalContext();
                                           ^
```

Looks like getGlobalContext has been removed form the public API in version 3.9 of LLVM.

Upstream ref: https://reviews.llvm.org/D19094
Workaround suggestion: https://reviews.llvm.org/D19094#417484
Fix used in the LLVM code base: https://reviews.llvm.org/rL266379

Basically the suggestions seems to be, don't use a global context. Makes sense for future work parallelization.

```
In file included from /home/u/Desktop/revng_play/revamb/main.cpp:35:0:
/home/u/Desktop/revng_play/revamb/ptcinterface.h:14:17: fatal error: ptc.h: No such file or directory
 #include "ptc.h"
                 ^
compilation terminated.
```

It looks like this may be resolved once your patches have been merged with Qemu; https://lists.gnu.org/archive/html/qemu-devel/2016-11/msg05477.html

I realize most of these issue may be related to me running LLVM v3.9 instead of v3.8. Is rev.ng intended to work with an unmodified version of LLVM or should the modified version at https://github.com/revng/llvm be used?

Cheers /u

**Comment by aleclearmind** [comment]
Currently the only supported method of building rev.ng is using [orchestra](https://github.com/revng/orchestra).

We're using a (very slightly) customized version of LLVM 3.8 ([revng/llvm](https://github.com/revng/llvm)). Eventually we'll move to 4.0.

We're also using a custom version of QEMU, the patches we're trying to upstream are on a more recent version of QEMU, and once they're upstream we'll update revamb to use them. For now, use the customized QEMU in [revng/qemu](https://github.com/revng/qemu).

But, seriously, orchestra will do everything for you! Don't try it yourself.

**Comment by mewmew** [comment]
> Currently the only supported method of building rev.ng is using orchestra.

Ok, I'll look into getting orchestra to compile revamb and its dependencies. I ran into some problems using it, but will open a dedicated issue for those.

Thanks for the feedback on the intended future support for LLVM and QEMU.

> Eventually we'll move to 4.0.

Does that mean vanilla LLVM 4.0, or a customized version?

> We're also using a custom version of QEMU, the patches we're trying to upstream are on a more recent version of QEMU, and once they're upstream we'll update revamb to use them.

That's great to hear!

**Comment by aleclearmind** [comment]
I hope we'll be able to use vanilla LLVM.
